### PR TITLE
feat(filter): add HTTP Basic Authentication filter (RFC 7617)

### LINKED ISSUE #557: HTTP Basic Auth filter
## What

HTTP Basic Authentication filter for downstream client requests.

## Why

Basic Auth is the simplest auth mechanism and is still used in internal/dev environments and legacy integrations. It complements the existing API key and JWT auth filters.

## Goals

- Extract credentials from `Authorization: Basic` header
- Validate against configurable credential store (inline, KV store)
- 401 with `WWW-Authenticate: Basic` on failure

### PR DESCRIPTION
**Pull request by eoinfennessy**
Closes https://github.com/praxis-proxy/praxis/issues/557

## Summary
- Add a new `basic_auth` HTTP filter that validates credentials from the `Authorization: Basic` header against inline username/password pairs or a runtime KV store
- Uses constant-time SHA-256 password comparison (via `subtle`) with dummy hashing for unknown users to prevent timing-based enumeration
- Supports configurable realm, authorization header stripping, and env var password resolution
- Includes unit tests, integration tests, example config, and filter reference docs

## Test plan
- [x] `cargo test -p praxis-proxy-filter` — unit tests for config parsing, credential validation, and edge cases
- [x] `cargo test -p praxis-tests-integration --test suite -- basic_auth` — integration test exercising the example config end-to-end
- [x] `make lint` — clippy and formatting checks pass
- [x] Manual verification with example config: `cargo run -p praxis-proxy -- -c examples/configs/security/basic-auth.yaml` and test with `curl -u admin:secret`, `curl` without credentials, and `curl -u admin:wrong`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by eoinfennessy** [comment]
> PR too large: 964 lines added (limit: 750, excludes Cargo files, tests, docs, examples, and benchmarks). Please split into smaller PRs. Add skip/pr-conventions label to override.

I don't think it makes sense to split this PR - it introduces a single filter, and splitting its implementation into multiple partially functioning PRs may make the review process more difficult

File: filter/src/builtins/http/security/basic_auth/config.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    type Error = String;
+
+    fn try_from(raw: RawBasicAuthConfig) -> Result<Self, Self::Error> {
+        let source = match (raw.credentials.is_empty(), raw.kv_store) {
+            (false, Some(_)) => return Err("both 'credentials' and 'kv_store' are set (use exactly one)".into()),
+            (true, None) => return Err("one of 'credentials' or 'kv_store' must be set".into()),
+            (false, None) => CredentialSourceConfig::Inline(raw.credentials),
+            (true, Some(store)) => CredentialSourceConfig::KvStore(store),
+        };
+        Ok(Self {
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply filter/src/builtins/http/security/basic_auth/config.rs:76]: Done in [8de307b](https://github.com/praxis-proxy/praxis/pull/824/commits/8de307bf7bb98956436b7bb169988214b5019b40)

File: filter/src/builtins/http/security/basic_auth/config.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                    raw.username
+                ));
+            },
+            (None, None) => {
+                return Err(format!(
+                    "user '{}' must have either 'password' or 'env_var'",
+                    raw.username
+                ));
+            },
+        };
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply filter/src/builtins/http/security/basic_auth/config.rs:159]: Done in [6b2ff47](https://github.com/praxis-proxy/praxis/pull/824/commits/6b2ff47b7c2ddb7e57ed99c604c6eb3f21b69da8)

File: docs/filters/http/security/basic_auth.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+# `basic_auth`
+
+HTTP Basic Authentication filter (RFC 7617).
+
+## Configuration Notes
+
+Extracts credentials from the `Authorization: Basic` header, validates against a configurable credential source (inline list or runtime KV store), and returns 401 with `WWW-Authenticate: Basic realm="..."` on failure.
+
+let yaml: serde_yaml::Value = serde_yaml::from_str( r#" credentials: - username: admin password: secret - username: deploy env_var: DEPLOY_PASSWORD "#, ) .unwrap(); let filter = BasicAuthFilter::from_config(&yaml).unwrap(); assert_eq!(filter.name(), "basic_auth");
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply docs/filters/http/security/basic_auth.md]: Fixed in [36a7041](https://github.com/praxis-proxy/praxis/pull/824/commits/36a704163f23d25e7a0569ac98babacd575e45a2). I removed the doctest and added a YAML example following the pattern of other filters.

File: tests/integration/tests/suite/examples/security_examples.rs [hunk 4]
```diff
@@ -232,3 +232,72 @@ fn downstream_read_timeout_normal_request() {
...
+    let raw = http_send(
+        proxy.addr(),
+        "GET / HTTP/1.1\r\n\
+         Host: localhost\r\n\
+         Authorization: Basic YWRtaW46c2VjcmV0\r\n\
+         Connection: close\r\n\r\n",
+    );
+    assert_eq!(parse_status(&raw), 200, "valid credentials should return 200");
+    assert_eq!(parse_body(&raw), "protected", "response should come from backend");
+}
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply tests/integration/tests/suite/examples/security_examples.rs:283]: Added test verifying `Authorization` header is stripped in [21d1611](https://github.com/praxis-proxy/praxis/pull/824/commits/21d16115533dea49c1170ecb8f5b767c92455ad8)

File: tests/integration/tests/suite/examples/security_examples.rs [hunk 5]
```diff
@@ -232,3 +232,72 @@ fn downstream_read_timeout_normal_request() {
...
+    // "admin:wrong" base64-encoded
+    let raw = http_send(
+        proxy.addr(),
+        "GET / HTTP/1.1\r\n\
+         Host: localhost\r\n\
+         Authorization: Basic YWRtaW46d3Jvbmc=\r\n\
+         Connection: close\r\n\r\n",
+    );
+    assert_eq!(parse_status(&raw), 401, "wrong password should return 401");
+}
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply tests/integration/tests/suite/examples/security_examples.rs:362]: Added a test case for this in [95b4c66](https://github.com/praxis-proxy/praxis/pull/824/commits/95b4c66735e7ce4f606ada028a942c14b7036517)

File: filter/src/builtins/http/security/basic_auth/filter.rs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+/// users to prevent timing-based user enumeration.
+///
+/// Both inputs are `[u8; 32]` SHA-256 digests, so `ct_eq` always
+/// performs a full 32-byte comparison without short-circuiting on
+/// length mismatch.
+fn verify_password_hash(provided: &[u8; 32], stored: Option<&[u8; 32]>) -> bool {
+    const DUMMY_HASH: [u8; 32] = [0_u8; 32];
+    let expected = stored.unwrap_or(&DUMMY_HASH);
+    let matches: bool = provided.ct_eq(expected).into();
+    matches && stored.is_some()
```
Comment (alexsnaps) [inline filter/src/builtins/http/security/basic_auth/filter.rs:211]: Not a huge deal, but the `.ct_eq` is useless when stored is `Option::None`.

File: filter/src/builtins/http/security/basic_auth/filter.rs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        .ok()
+        .and_then(|bytes| String::from_utf8(bytes).ok())?;
+
+    // RFC 7617: split on first colon; password may contain colons.
+    let (username, password) = decoded.split_once(':')?;
+    Some((username.to_owned(), password.to_owned()))
+}
+
+/// Dispatches credential verification to the configured source.
+fn verify_credentials(source: &CredentialSource, ctx: &HttpFilterContext<'_>, username: &str, password: &str) -> bool {
```
Comment (alexsnaps) [inline filter/src/builtins/http/security/basic_auth/filter.rs]: Can we make all these methods (e.g. here move to `impl CredentialSource`) - which they effectively are? Makes for more idiomatic Rust code and easier on refactoring (moves, which move the impl with the type).

**Review by alexsnaps — COMMENTED** [review]
Thanks for the PR @eoinfennessy 

I'll be transparent here, while this does add basic HTTP auth, I don't know how much of the implementation is currently appropriate. We have on-going efforts to [implement auth backed by CPEX](https://github.com/praxis-proxy/demos/blob/main/demos/cpex/cpex-cel.yaml). But that'd put the bar real high to get someone going with Basic HTTP Auth. Which, in turn, is possibly desirable, as would we really recommend using Basic Auth RFC 7235 ? Would any one?

Anyways, as you can see this is somewhat conflicting within me:
 - This PR adds an easy to use basic auth filter;
 - I like that it can leverage a KV store (and not plain text passwords in a file);
 - I know CPEX should be able to provide a "completed" story to auth.

With all that being said, this is fairly lightweight too. Wondering if we should make add this behind some (broader) feature flag... like _that's easy, but don't you use this unless you really need to_ feature flag... 

What's your take?

File: filter/src/builtins/http/security/basic_auth/filter.rs (hunk 6 above)
Comment (alexsnaps) [reply filter/src/builtins/http/security/basic_auth/filter.rs:211]: >Uses constant-time SHA-256 password comparison (via subtle) with dummy hashing for unknown users to prevent timing-based enumeration

🤦

File: docs/filters/http/security/basic_auth.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+## Example
+
+```yaml
+filter: basic_auth
+realm: "Restricted"
+strip_authorization: true
+credentials:
+  - username: admin
+    password: secret
```
Comment (alexsnaps) [inline docs/filters/http/security/basic_auth.md:22]: Should we support digests straight here? Along (over?) plain text passwords?

**Comment by eoinfennessy** [comment]
Hey @alexsnaps,

Thanks for reviewing this. I appreciate the transparency around usefulness of a Basic Auth filter.

My take is that Basic Auth is an outdated / legacy authn mechanism that is only really useful for quickstart / dev / demo purposes. I think if we are introducing a Basic Auth filter, being honest about its limitations is important.

Re. password storage: we _could_ store passwords as SHA-1 digests (like Envoy's impl.), or SHA-2. But this only really gives an illusion of security - without salting or key stretching these algos offer limited protection, and are not much better than plaintext password storage for an internet-exposed gateway. bcrypt or argon2 could be used for secure password digests suitable for storage, and they could be expressed as password hashes using Modular Crypt Format. But these algos are not FIPS-approved. Maybe this introduces complexity and maintenance costs to an auth mechanism that probably should not be recommended for production use?

The way I see it, we have three options:
1. close this PR in favour of proper auth mechanisms (tradeoff: no quickstart option)
2. go with what is implemented here - plaintext password storage is obviously/explicitly insecure (tradeoff: no production value whatsoever)
3. make this impl. more secure by using bcrypt for stored password digests (tradeoff: maintenance costs for an auth mechanism that probably should no longer be used; more work required hashing passwords for quickstart use cases)

I'm not strongly against any of these options. I kind of feel like option 2 is the most honest, and could be accompanied by a clear warning log message stating that this filter is not suitable for production use. And/or gated behind some sort of "insecure" feature flag.

WDYT? I'm very happy to go with whatever you feel is right for the project.

File: filter/src/builtins/http/security/basic_auth/filter.rs (hunk 7 above)
Comment (eoinfennessy) [reply filter/src/builtins/http/security/basic_auth/filter.rs]: Thanks for the tip! Done in [da5d08e](https://github.com/praxis-proxy/praxis/pull/824/commits/da5d08ec4b6d4b71bdd281ec47297bca59ac0071)

File: docs/filters/http/security/basic_auth.md (hunk 8 above)
Comment (eoinfennessy) [reply docs/filters/http/security/basic_auth.md:22]: I shared thoughts on this in https://github.com/praxis-proxy/praxis/pull/824#issuecomment-5052732902

**Comment by alexsnaps** [comment]
I share pretty much all of you analysis. My proposal towards the end was "let's go for option 2: go with what is implemented here" _but behind a feature flag_, wdyt? 

Thinking maybe we should have possibly a "`-dev`" flavor build/binaries maybe, that'd have all these "ease of life/getting started" features enabled, but not recommended for production? 

Any @praxis-proxy/devs have other opinions on this?

**Comment by franciscojavierarceo** [comment]
>Thinking maybe we should have possibly a "-dev" flavor build/binaries maybe, that'd have all these "ease of life/getting started" features enabled, but not recommended for production?
>
>Any @praxis-proxy/devs have other opinions on this?

I don't think @leseb likes dev flags but I am fine with way to configure this so long as it's disabled by default.

**Comment by alexsnaps** [comment]
> so long as it's disable by default

Right that was my thinking, it's not a (set of) filters available by default. So disabled by default. 
But were you thinking of a runtime flag? 🤔 I was thinking of a build time one. Tho possibly having a "developer friendly" build, which then would have these enabled... maybe it's bad idea. But I can see how requiring people to build the with these features would be annoying to some as well...

**Review by shaneutt — COMMENTED** [review]
Appreciate the effort here @eoinfennessy! 🙇 

It occurs to me that given #12 and all the design work that needs to be done therein, this patch could end up getting significantly changed in the near future.

I think getting it in now so we have something in place, recognizing that it might get re-hauled soon is fine, but we should flag it in some way that makes it clear to the reader. I suggest putting it behind an build flag and have it off by default and labeled experimental as one way to communicate that.

**Comment by eoinfennessy** [comment]
Thanks for sharing your thoughts on this folks.

Commit [9faa73a](https://github.com/praxis-proxy/praxis/pull/824/commits/9faa73a40681190ee1916d56220cb4708e4606eb) gates this filter behind a `basic-auth-filter` Cargo feature. I also introduced a `dev` feature to be composed of all dev-specific features including this basic auth filter. The commit also adds a warning log whenever the build contains any experimental features. Another warning is logged when the proxy is configured to use the basic auth filter. I also added a `build-dev` target to the Makefile to build with the `dev` feature.

File: filter/src/builtins/http/security/basic_auth/config.rs [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+            (false, Some(_)) => return Err("both 'credentials' and 'kv_store' are set (use exactly one)".into()),
+            (true, None) => return Err("one of 'credentials' or 'kv_store' must be set".into()),
+            (false, None) => CredentialSourceConfig::Inline(raw.credentials),
+            (true, Some(store)) => CredentialSourceConfig::KvStore(store),
+        };
+        if raw.realm.contains('"') || raw.realm.bytes().any(|b| b < 0x20 || b == 0x7F) {
+            return Err("realm must not contain double-quotes or control characters".into());
+        }
+
+        Ok(Self {
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply filter/src/builtins/http/security/basic_auth/config.rs:76]: Addressed in 81f2074 — added `\\` to the rejected character set and a `rejects_realm_with_backslash` unit test.

File: filter/src/builtins/http/security/basic_auth/tests.rs [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+async fn rejects_missing_authorization_header() {
+    let f = make_filter(&[("admin", "secret")], "TestRealm");
+    let req = crate::test_utils::make_request(http::Method::GET, "/");
+    let mut ctx = crate::test_utils::make_filter_context(&req);
+
+    let action = f.on_request(&mut ctx).await.unwrap();
+    assert_rejection_with_challenge(&action, "TestRealm");
+}
+
+#[tokio::test]
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply filter/src/builtins/http/security/basic_auth/tests.rs:229]: Addressed in b12e51e — added `rejects_non_utf8_base64` and `rejects_credentials_without_colon` unit tests.

File: Makefile [hunk 11]
```diff
@@ -103,6 +103,9 @@ build:
 	cargo build --workspace
 	cargo build --workspace --benches
```
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply Makefile:105]: Pushing back on this. Currently, the Makefile does not test any optional features. Creating a testing strategy for features is definitely worth doing, but seems outside the scope of this PR.

[the remark this answers is not in the document]
Comment (shaneutt) [reply Makefile:105]: Please create a follow-up issue to take care of this if you have some time, then feel free to mark it resolved :vulcan_salute:

File: docs/filters/http/security/basic_auth.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
+<!-- Generated by: cargo xtask generate-filter-docs -->
+<!-- Do not edit manually -->
+
+# `basic_auth`
```
Comment (shaneutt) [inline docs/filters/http/security/basic_auth.md:4]: I think we have marked filters as experimental before, there's some precedent for that... 🤔

**Review by shaneutt — APPROVED** [review]
This is marked as experimental, it's nice and contained in it's own section, so I think we're good to go for now. Once all existing comments are resolved, LGTM 👍

File: Makefile (hunk 11 above)
[the remark this answers is not in the document]
Comment (eoinfennessy) [reply Makefile:105]: Thanks, I created https://github.com/praxis-proxy/praxis/issues/878 to track this

File: docs/filters/http/security/basic_auth.md (hunk 12 above)
Comment (eoinfennessy) [reply docs/filters/http/security/basic_auth.md:4]: I made it clearer in the generated docs that this is experimental, following the pattern of the policy filter. [48e9a26](https://github.com/praxis-proxy/praxis/pull/824/commits/48e9a269138d13d7597d13e349fb31c7f2665ad4)
